modules/dxp/apps/commerce-avalara/commerce-avalara-connector/src/main/java/com/liferay/commerce/avalara/connector/internal/helper/CommerceAvalaraConnectorHelperImpl.java

```diff
@@ -0,0 +1,258 @@
+import com.liferay.commerce.tax.engine.fixed.model.CommerceTaxFixedRateAddressRel;
+import com.liferay.commerce.tax.engine.fixed.service.CommerceTaxFixedRateAddressRelLocalService;
+import com.liferay.commerce.tax.model.CommerceTaxMethod;
+import com.liferay.commerce.tax.service.CommerceTaxMethodLocalService;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.language.LanguageUtil;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.util.HashMapBuilder;
+import com.liferay.portal.kernel.util.LocaleUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.Validator;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import net.avalara.avatax.rest.client.models.TaxCodeModel;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Riccardo Alberti
+ * @author Katie Nesterovich
+ */
+@Component(
+	enabled = false, immediate = true,
+	service = CommerceAvalaraConnectorHelper.class
+)
+public class CommerceAvalaraConnectorHelperImpl
+	implements CommerceAvalaraConnectorHelper {
+
+	@Override
+	public void createTaxCategories(long userId) throws Exception {
+		List<TaxCodeModel> taxCodeModels =
+			_commerceAvalaraConnector.listTaxCodes();
+
+		ServiceContext serviceContext = new ServiceContext();
+
+		serviceContext.setUserId(userId);
+
+		for (TaxCodeModel taxCodeModel : taxCodeModels) {
+			Map<Locale, String> nameMap = Collections.singletonMap(
+				LocaleUtil.getSiteDefault(), taxCodeModel.getDescription());
+
+			_cpTaxCategoryLocalService.addCPTaxCategory(
+				taxCodeModel.getTaxCode(), nameMap, Collections.emptyMap(),
+				serviceContext);
+		}
+	}
+
+	@Override
+	public void removeByAddressEntries(CommerceTaxMethod commerceTaxMethod) {
+		_commerceTaxFixedRateAddressRelLocalService.
+			deleteCommerceTaxFixedRateAddressRelsByCommerceTaxMethodId(
+				commerceTaxMethod.getCommerceTaxMethodId());
+	}
+
+	@Override
+	public void updateByAddressEntries(long groupId) throws Exception {
+		String taxRateByZipCode =
+			_commerceAvalaraConnector.getTaxRateByZipCode();
+
+		if (Validator.isBlank(taxRateByZipCode)) {
+			return;
+		}
+
+		CommerceTaxMethod commerceTaxMethod =
+			_commerceTaxMethodLocalService.fetchCommerceTaxMethod(
+				groupId, CommerceAvalaraConstants.KEY);
+
+		if (commerceTaxMethod == null) {
+			return;
+		}
+
+		String[] taxRatesByZipCodeLines = StringUtil.splitLines(
+			taxRateByZipCode);
+
+		CommerceCountry commerceCountryUS =
+			_commerceCountryLocalService.fetchCommerceCountry(
+				commerceTaxMethod.getCompanyId(), "US");
+
+		if (commerceCountryUS == null) {
+			commerceCountryUS = _addUnitedStates(commerceTaxMethod);
```

Comment:
Not sure if this is necessary, maybe just throw an exception?

Reply:
In our setup guide we will advise the admin to run all the necessary health checks before they reach this stage, so ideally this check will never be true. However, we still want this process to execute if they haven't run the health checks (per earlier conversation with Corbin).